lambda/dissect_letrec.ml

```diff
@@ -0,0 +1,592 @@
+     that other variables can't be refered to by the expression,
+     but if it happens, then the value is useless.
+
+   We consider two cases for expressions that we consider of
+   known size (the [Static] case in [is_valid_recursive_expression]):
+   the makeblock primitive and function declarations. Makeblocks will
+   be preallocated, while all the functions will be turned into a single
+   letrec.
+
+   The structure of the generated code will be:
+   {[
+     let c = const in
+     (* [consts]: constants *)
+     ...
+     let v = caml_alloc_dummy n in
+     (* [blocks]: With n the staticaly known size of blocks *)
+     ...
+     let p = expr in
+     (* [pre]: Values that do not depend on any other from the let-rec in
+        [Rec_check.Dereference] position *)
+     ...
+     let rec f x = ...
+     and g x = ...
+     and ...
+     in
+     (* [functions]: All the functions from the let-rec *)
+     caml_update_dummy v v_contents;
+     (* Initialisation ([effects]) *)
+     ...
+   ]}
+
+   Special care is taken to handle nested [let rec]s:
+   the recursive values of a letrec are often more than the ones
+   bound by the letrec expression itself. For instance
+
+   let rec f =
+     let rec g x = h (x+1)
+     and h x = i x
+     in
+     fun x -> g x
+   and i x =
+     if x > 33 then x
+     else f x
+
+   in this expression every function is recursively defined with
+   any other. Hence this is equivalent to
+
+   let rec f = fun x -> g x
+   and i x =
+     if x > 33 then x
+     else f x
+   and g x = h (x+1)
+   and h x = i x
+
+   However, there might be (a subset of) a local [let rec] that is indeed
+   internally recursive, but that is used by the top-level [let rec] in
+   [Dereference] positions. We carefully lift those within [pre], and handle
+   them recursively in the same pass to preserve the order of evaluation.
+
+   The analysis for which variables should remain in the inner [let rec], and
+   which are indeed part of the outer [let rec] is equivalent to the
+   [Rec_check.value_bindings] case.
+*)
+
+type block_type = Normal of int (* tag *) | Boxed_float
+
+type block = { block_type : block_type; size : int; }
+
+type letrec = {
+  blocks : (Ident.t * block) list;
+  (* Pre-allocated blocks.
+     Will result in [let id = caml_alloc_dummy size] or
+     [let id = caml_alloc_dummy_float size] *)
+  consts : (Ident.t * Lambda.structured_constant) list;
+  (* Statically known values *)
+  pre : tail:Lambda.lambda -> Lambda.lambda;
+  (* Prefix of the expression that does not depends on any recursive part.
+     This is presented as a function for easy 'concatenation':
+     to append 'expr': [fun ~tail -> Lsequence (expr, letrec.pre ~tail)] *)
+  effects : Lambda.lambda;
+  (* Effects that are applied afterwards. *)
+  functions : (Ident.t * Lambda.lfunction) list;
+  substitution : Ident.t Ident.Map.t;
+  (* Alias to recursive variables should be forbidden, to prevent
```

Comment:
A quick investigation of what would happen if we didn't substitute showed a problem of placement: an alias `let x = y in body` needs to be placed after the definition of `y` but before any expression that can refer to `x`. If `y` is not a function, putting the alias in `pre` should work, but if it is a function being recursively defined, then it would need to be defined recursively with the other functions, and this would either lose the property that the result of `dissect_letrec` only contains function definitions (except in the `pre` part, where it can contain independent let recs not dissected yet), or the aliases need to be eta-expanded, which would create unnecessary closures.
Anyway, next time we meet I'll be able to provide a better explanation of the substitution part.